### ISSUE TITLE
DatastoreException - Error Code/Reason are inconsistent when working with Emulator 

### ISSUE THREAD

**Issue by sai-pullabhotla** [GoogleCloudPlatform/google-cloud-java#2442]
I'm trying to implement https://github.com/sai-pullabhotla/catatumbo/issues/172 to throw specialized exceptions in certain cases, more specifically, for `NOT_FOUND` and `ALREADY_EXISTS` errors. 

As per this [page ](https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto)- 

```
NOT_FOUND = 5 
ALREADY_EXISTS = 6 

```
This works well and get the expected codes when working with the **real/live GCD**, but when using the **Datastore Emulator**, both test cases result in a code of `INVALID_ARGUMENT = 3`

FYI - Just did an update on gcloud/emulator and the issue remains. The gcloud versions are - 

```
Google Cloud SDK 171.0.0
beta 2017.03.24
bq 2.0.25
cloud-datastore-emulator 1.2.1
core 2017.09.11
gcloud
gsutil 4.27

```

**Comment by andreamlin** [comment]
In the constructor for DatastoreException
`public DatastoreException(int code, String message, String reason) `
the int `code` is meant to be an HTTP status code, e.g. 409 for "Already Exists" or 404 for "Not Found".

The Java Datastore library runs over HTTP 1.

**Comment by andreamlin** [comment]
Closing this issue for now.

**Comment by sai-pullabhotla** [comment]
@andreamlin - 

I don't think the analysis is correct. The error code in the `DatastoreException `does in fact equals to one of the codes defined in https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto. This is not HTTP status code. HTTP status codes are translated to one of these codes in `HttpDatastoreRpc#translate` method. Here is a simple test case for you to run. 

# Sample Code: 
```
import java.util.UUID;

import com.google.cloud.datastore.Datastore;
import com.google.cloud.datastore.DatastoreException;
import com.google.cloud.datastore.DatastoreOptions;
import com.google.cloud.datastore.Entity;
import com.google.cloud.datastore.KeyFactory;

public class DatastoreExceptionTest {

	public static void main(String[] args) {
		// Connect to real Datastore
		Datastore datastore = DatastoreOptions.getDefaultInstance().getService();

		// Connect to Emulator
		// Datastore datastore =
		// DatastoreOptions.newBuilder().setHost("localhost:9999").build().getService();
		KeyFactory keyFactory = datastore.newKeyFactory();
		Entity entity = Entity.newBuilder(keyFactory.setKind(UUID.randomUUID().toString()).newKey(1000L)).build();
		try {
			datastore.update(entity);
		} catch (DatastoreException e) {
			System.out.printf("Code: %s%nReason: %s%nMessage: %s%n", e.getCode(), e.getReason(), e.getMessage());
		}
	}
}

```
Code tries to update a nonexistent entity, so a NOT_FOUND exception is expected. 

# Output when running against GCD: 

```
Code: 5
Reason: NOT_FOUND
Message: no entity to update: app: "s~my-project-id"
path <
  Element {
    type: "ffc7fba1-1169-485c-b48e-6dd6ad70e8e5"
    id: 1000
  }
>
```

# Output when running the test against Emulator 

```
Code: 3
Reason: INVALID_ARGUMENT
Message: no entity to update

```

**Comment by andreamlin** [comment]
The code sample is really helpful, thanks!

It looks like the discrepancy is due to different implementations of the emulator vs the actual Datastore backend.  I opened an internal bug #67105935 for the Datastore team to address.

**Comment by wsh** [comment]
Working on this (on the Datastore team).

**Comment by wsh** [comment]
Hey there @sai-pullabhotla! This has been fixed with the latest emulator release, version 1.4.0. This change hasn't been released to the gcloud SDK component (it should be there next week) but, if you'd like a preview, you can [download the emulator stand-alone from here](https://storage.googleapis.com/gcd/tools/cloud-datastore-emulator-1.4.0.zip). Thanks for your patience and help here. :smile:

**Comment by sai-pullabhotla** [comment]
@wsh - Thanks for the update. I will check it out and let you know.

**Comment by sai-pullabhotla** [comment]
@wsh  - I've verified this and here is what I found:

After upgrading my locally installed **gcloud CLI**, the issue is fixed. However, the builds on Travis CI are failing as I use **LocalDatastoreHelper** to run the tests. The LocalDataStoreHelper is still using version **1.2.0** of the DatastoreEmulator, where as the latest is **1.4.0**. 

Is there a reason why the LocalDatastoreHelper is set to download 1.2.0, rather than the latest? Is it possible to update the LocalDatastoreHelper to download the latest version?
